pkg/builder/in_memory_build_queue.go

```diff
@@ -0,0 +1,1489 @@
+
+// GetCapabilities returns the Remote Execution protocol capabilities
+// that this service supports.
+func (bq *InMemoryBuildQueue) GetCapabilities(ctx context.Context, in *remoteexecution.GetCapabilitiesRequest) (*remoteexecution.ServerCapabilities, error) {
+	return &remoteexecution.ServerCapabilities{
+		CacheCapabilities: &remoteexecution.CacheCapabilities{
+			DigestFunction: util.SupportedDigestFunctions,
+			ActionCacheUpdateCapabilities: &remoteexecution.ActionCacheUpdateCapabilities{
+				UpdateEnabled: false,
+			},
+			// CachePriorityCapabilities: Priorities not supported.
+			// MaxBatchTotalSize: Not used by Bazel yet.
+			SymlinkAbsolutePathStrategy: remoteexecution.SymlinkAbsolutePathStrategy_ALLOWED,
+		},
+		ExecutionCapabilities: &remoteexecution.ExecutionCapabilities{
+			DigestFunction: remoteexecution.DigestFunction_SHA256,
+			ExecEnabled:    true,
+			ExecutionPriorityCapabilities: &remoteexecution.PriorityCapabilities{
+				Priorities: []*remoteexecution.PriorityCapabilities_PriorityRange{
+					{MinPriority: math.MinInt32, MaxPriority: math.MaxInt32},
+				},
+			},
+		},
+		// TODO: DeprecatedApiVersion.
+		LowApiVersion:  &semver.SemVer{Major: 2},
+		HighApiVersion: &semver.SemVer{Major: 2},
+	}, nil
+}
+
+// Execute an action by scheduling it in the build queue. This call
+// blocks until the action is completed.
+func (bq *InMemoryBuildQueue) Execute(in *remoteexecution.ExecuteRequest, out remoteexecution.Execution_ExecuteServer) error {
+	// Fetch the action and command corresponding to the execute
+	// request. Ideally, a scheduler is be oblivious of what these
+	// look like, if it weren't for the fact that Action.DoNotCache
+	// and Command.Platform are used for scheduling decisions.
+	//
+	// To prevent loading these messages from the Content
+	// Addressable Storage (CAS) multiple times, the scheduler holds
+	// on to them and passes them on to the workers.
+	ctx := out.Context()
+	actionDigest, err := util.NewDigest(in.InstanceName, in.ActionDigest)
+	if err != nil {
+		return util.StatusWrap(err, "Failed to extract digest for action")
+	}
+	action, err := bq.contentAddressableStorage.GetAction(ctx, actionDigest)
+	if err != nil {
+		return util.StatusWrap(err, "Failed to obtain action")
+	}
+	commandDigest, err := actionDigest.NewDerivedDigest(action.CommandDigest)
+	if err != nil {
+		return util.StatusWrap(err, "Failed to extract digest for command")
+	}
+	command, err := bq.contentAddressableStorage.GetCommand(ctx, commandDigest)
+	if err != nil {
+		return util.StatusWrap(err, "Failed to obtain command")
+	}
+	platformKey, err := newPlatformKey(in.InstanceName, command.Platform)
+	if err != nil {
+		return err
+	}
+
+	bq.enter(bq.clock.Now())
+	defer bq.leave()
+
+	pq, ok := bq.platformQueues[platformKey]
+	if !ok {
+		return status.Errorf(codes.FailedPrecondition, "No workers exist for instance %#v platform %s", platformKey.instance, platformKey.platform)
+	}
```

Comment:
The implementation of the platform logic is well done, but matches exact platform properties only. One suggested RW API, [API Design: Remote Workers API](https://docs.google.com/document/d/1s_AzRRD2mdyktKUj2HWBn99rMg_3tcPvdjx3MPbFidU/edit#heading=h.1lt0k9rr0yu4) under the chapter _Configurations_, suggests possibility to run on any worker that fulfills **at least** the properties advertised by the worker. One example is that one action requires SSE4 but another one does not, but both can run on the same worker. I have also seen a good use of advertising for example `DockerImage=` from the worker side, to support a worker that can accept actions for any image.

Changing this logic is a bit cumbersome and can be done in a separate pull request. I'm thinking in letting `Execute()` populate the `platformQueues` map and let workers poll multiple `platformQueues` in `Synchronize()`.

Reply:
You got that right! The first step was to implement just the basic ways of matching platforms. Subsequent improvements to add fuzzy ways of matching platform queues should be added later on.

The question becomes what an efficient lookup structure looks like, especially when the number of platform queues becomes large. Right now we simply stringify the platform properties and do a map lookup, which should scale well enough.

Doing subset/superset matching by default seems a bit risky in my case, as it means the matching criteria may change even if seemingly unrelated platform queues get created/removed from the system. But that's not a big problem if we get the UX right.